src/hotspot/share/prims/jvmtiRawMonitor.cpp

```diff
@@ -410,38 +375,45 @@ int JvmtiRawMonitor::raw_wait(jlong millis, Thread* self) {
   OrderAccess::fence();
 
   intptr_t save = _recursions;
-  _recursions = 0;
-  _waiters++;
-  ret = simple_wait(self, millis);
-  _recursions = save;
-  _waiters--;
-
-  guarantee(self == _owner, "invariant");
 
   if (self->is_Java_thread()) {
     JavaThread* jt = self->as_Java_thread();
     guarantee(jt->thread_state() == _thread_in_native, "invariant");
-    for (;;) {
-      if (!SafepointMechanism::should_process(jt)) {
-        // Not suspended so we're done here:
-        break;
+
+    _recursions = 0;
+    _waiters++;
+    ret = simple_wait(self, millis);
+    _waiters--;
+
+    {
+      ThreadInVMfromNative tivmfn(jt);
+      for (;;) {
+        ExitOnSuspend eos(this);
+        {
+          ThreadBlockInVMPreprocess<ExitOnSuspend> tbivmp(jt, eos);
+          simple_enter(jt);
+          _recursions = save;
```

Comment:
I think you should restore `_recursions` only when finally leaving the loop because if you do it here and exit the monitor again because of a suspend then another thread could hit the `guarantee(_recursions == 0, "invariant")` at L349. This is a preexisting issue but since you're modifying this I reckon it makes sense to correct it.

Reply:
Fixed